vi-start-to-finish.py

```diff
@@ -0,0 +1,223 @@
+
+# these are much shorter tasks, so we start a new cluster:
+cluster.close()
+cluster = Cluster(walltime='01:00:00', cores=1, memory='4gb', job_cpu=1, job_mem='4gb')
+cluster.adapt(minimum=0, maximum=48, target_duration='4h')
+
+
+intermediate_images = sorted(glob.glob('*/*.illum.png'))
+exp = (r'(?P<dir>.*)/(?P<plate>mfgtmp_\d*)_'
+       r'(?P<well>[A-P]\d\d)f(?P<field>\d\d)d(?P<channel>\d).illum.png')
+
+def plate_well(fn):
+    match = re.match(exp, fn)
+    if match is not None:
+        return match['plate'], match['well']
+    else:
+        return 'none', 'none'
+
+def ch(fn):
+    match = re.match(exp, fn)
+    if match is not None:
+        return int(match['channel'])
+    else:
+        return 'none'
+
+by_coord = tz.groupby(plate_well, intermediate_images)
+
+order = generate_spiral((5, 5), 'right', clockwise=False) 
+
+def montage(coord, fns):
+    plate, well = coord
+    fn_out = os.path.join(OUT, plate, '-'.join(coord)) + '.jpg'
+    # uncomment line below if starting pipeline halfway
+    #if os.path.exists(fn_out):
+    #    return
+    by_ch = tz.groupby(ch, fns)
+    montaged = {}
+    for c, fns_ch in sorted(by_ch.items()):
+        montaged[c] = pre.montage_with_missing(fns_ch, order=order,
+                                                re_string=exp,
+                                                re_group='field')[0]
+    stacked = pre.stack_channels(montaged.values(), order=[None, 1, 0])
+    io.imsave(fn_out, stacked, quality=95)
+
+
+plates = set(coord[0] for coord in by_coord)
+for plate in plates:
+    os.makedirs(os.path.join(OUT, plate), exist_ok=True)
+
+
+client = Client(cluster)
+results = [client.submit(montage, coord, fns)
+           for coord, fns in by_coord.items()]
+
+for r in tqdm(as_completed(results), desc='montage', total=len(by_coord)):
+    r.cancel()
+
+t3 = time.time()
+print(f'montaged in {ftime(t3 - t2)}')
+
+
+
+
+
+############### feature computation #####################
+
+os.chdir(OUT)
+
+plates = [d for d in os.listdir('.')
+          if os.path.isdir(d) and d.startswith('mfgtmp')]
+
+rows = [chr(n + ord('A')) for n in range(16)]
+cols = [f'{i:02}' for i in range(1, 25)]
+wells = [''.join([row, col]) for row in rows for col in cols]
+
+all_plate_wells = list(itertools.product(plates, wells))
+all_filenames = [os.path.join(plate, '-'.join([plate, well]) + '.jpg')
+                 for plate, well in all_plate_wells]
+plate_wells, filenames = zip(*[(plate_well, filename)
+                               for plate_well, filename
+                               in zip(all_plate_wells, all_filenames)
+                               if os.path.exists(filename)])
+
+
+def features_with_names(fn):
+    image = io.imread(fn)
+    fvector, names = feat.default_feature_map(image, channels=[2, 1],
+                                              channel_names=['dapi', 'polyA'],
+                                              sample_size=1000)
+    return fvector, names
+
+
+def features(num_fn):
```

Comment:
I don't really understand why you need this wrapper function around `features_with_names`

Reply:
Added a comment to clarify.